connectors/sources/microsoft_teams.py

```diff
@@ -186,25 +366,59 @@ def set_logger(self, logger_):
     async def fetch(self, url):
         return await self._get_json(absolute_url=url)
 
+    async def pipe(self, url, stream):
+        async for response in self._get(absolute_url=url, is_token=False):
+            async for data in response.content.iter_chunked(FILE_WRITE_CHUNK_SIZE):
+                await stream.write(data)
+
+    async def scroll(self, url):
+        scroll_url = url
+
+        while True:
+            graph_data = await self._get_json(scroll_url)
+            # We're yielding the whole page here, not one item
+            yield graph_data.get("value", [])
+
+            if not graph_data.get("@odata.nextLink"):
+                break
+            scroll_url = graph_data.get("@odata.nextLink")
+
     async def _get_json(self, absolute_url):
-        response = await anext(self._get(absolute_url=absolute_url))
-        return await response.json()
+        async for response in self._get(absolute_url=absolute_url):
+            return await response.json()
 
     @retryable(
         retries=RETRY_COUNT,
         interval=RETRY_INTERVAL,
         strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
         skipped_exceptions=[NotFound, PermissionsMissing],
     )
-    async def _get(self, absolute_url):
+    async def _get(self, absolute_url, is_token=True):
         try:
-            token = await self._api_token.get_with_username_password()
-            self._logger.debug(f"Calling Microsoft Teams Endpoint: {absolute_url}")
-            async with self._http_session.get(
-                url=absolute_url,
-                headers={"authorization": f"Bearer {token}"},
-            ) as resp:
-                yield resp
+            if is_token:
+                if any(
+                    substring in absolute_url
+                    for substring in [
+                        "/drive",
+                        "/items/root/children",
+                        "/children?$filter=name eq",
+                        "/events",
+                    ]
+                ):
+                    token = await self._api_token.get_with_client()
+                else:
+                    token = await self._api_token.get_with_username_password()
```

Comment:
This could be abstracted into a token-generating func to make the `_get`  func cleaner.

Reply:
But in that case we need to add this condition at 3 places. In `get_with_client` method we need to add one condition for checking cached_value, one for fetch_token and one for setting token to cache.
To remove it from here we can include one more fucntion where it can decide which function needs to call for generating token `get_with_client` or `get_with_username_password` what do you think on this?

Reply:
If changing this has such an impact, let's leave it where it is. The token isn't needed anywhere so abstraction is just a preference, not a requirement.